mlflow/fastai.py

```diff
@@ -0,0 +1,192 @@
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    with open(os.path.join(path, conda_env_subpath), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.fastai",
+                        data=model_data_subpath, env=conda_env_subpath)
+    mlflow_model.add_flavor(FLAVOR_NAME, fastai_version=fastai.__version__, data=model_data_subpath)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def log_model(fastai_learner, artifact_path, conda_env=None, registered_model_name=None, **kwargs):
+    """
+    Log a fastai model as an MLflow artifact for the current run.
+
+    :param fastai_learner: Fastai model (an instance of `fastai.Learner`_) to be saved.
+    :param artifact_path: Run-relative artifact path.
+    :param conda_env: Either a dictionary representation of a Conda environment or the path to a
+                      Conda environment yaml file. If provided, this describes the environment
+                      this model should be run in. At minimum, it should specify the dependencies
+                      contained in :func:`get_default_conda_env()`. If ``None``, the default
+                      :func:`get_default_conda_env()` environment is added to the model.
+                      The following is an *example* dictionary representation of a Conda
+                      environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults'],
+                            'dependencies': [
+                                'python=3.7.0',
+                                'fastai=1.0.60',
+                            ]
+                        }
+    :param registered_model_name: Note:: Experimental: This argument may change or be removed in a
+                                  future release without warning. If given, create a model
+                                  version under ``registered_model_name``, also creating a
+                                  registered model if one with the given name does not exist.
+    :param kwargs: kwargs to pass to `fastai.Learner.export`_ method.
+    """
+    Model.log(artifact_path=artifact_path, flavor=mlflow.fastai,
+              registered_model_name=registered_model_name,
+              fastai_learner=fastai_learner, conda_env=conda_env, **kwargs)
+
+
+def _load_model(path):
+    from fastai.basic_train import load_learner
+    abspath = os.path.abspath(path)
+    path, file = os.path.split(abspath)
+    return load_learner(path, file)
+
+
+class _FastaiModelWrapper:
+    def __init__(self, learner):
+        self.learner = learner
+
+    def predict(self, dataframe):
+        test_data = TabularList.from_df(dataframe, cont_names=self.learner.data.cont_names)
+        self.learner.data.add_test(test_data)
+        return self.learner.get_preds(DatasetType.Test)[0].numpy()
```

Comment:
Thanks for opening a new PR! To continue the discussion from https://github.com/mlflow/mlflow/pull/2619#discussion_r406965943, maybe we ought to just return both the probabilities & labels given that that's what fastai returns / to reduce user confusion? I would imagine that for most use cases, a user would care about the labels rather than the probabilities, so seems nice to return the labels in addition to the probabilities for that reason as well .

 Also, do you know what happens in the regression case? Spent some time reading through the [fastai regression tutorial](https://docs.fast.ai/tutorial.data.html#A-regression-example), but couldn't quite find an example of using `get_preds` in the regression case - thanks!!

Reply:
For the regression case, it returns a tuple but the first tensor is the predictions and the second is the actual target (ground truth), so I think it's not useful for inference.

https://forums.fast.ai/t/tabular-regression-prediction-phase-returns-2-tensors/28158

The reason why I chose to return only probabilities is that the behavior changes between regression and classification as you can see.

Btw, I've created a small example for you to try:

https://gist.github.com/antoniomdk/1f2728ddcc32af5a715cd8a6a011015c

Reply:
Thanks! From what I understand (from adding a `help(learn.get_preds)` to the end of your example) it seems `get_preds` returns both predictions & targets (which are all 0 in the regression case) on the dataset:

```
get_preds(ds_type:fastai.basic_data.DatasetType=<DatasetType.Valid: 2>, activ:torch.nn.modules.module.Module=None, with_loss:bool=False, n_batch:Union[int, NoneType]=None, pbar:Union[fastprogress.fastprogress.ProgressBar, NoneType]=None) -> List[torch.Tensor] method of fastai.basic_train.Learner instance
    Return predictions and targets on `ds_type` dataset.
```

I wonder if we could actually return a Dataframe from this method containing `predictions` and `targets` columns, so that users can access whichever column they care about (e.g. directly get the targets if that's relevant for their use case, or get the raw predictions). Thoughts?

Reply:
We can always return a Dataframe with two columns because that's what ` learn.get_preds` returns. If we want to return always two columns we can do two things:

1. Return the data as it is
2. Return the data as it is for classification tasks and duplicate the first column for regression. In that way, the target vector contains predictions too.

IMO, the first option is better to maintain the consistency between the Learn method and the wrapper.

Reply:
:+1: option 1 (return the data as it is) sounds good to me!

Reply:
I agree. I'll work on that during this weekend.

Reply:
I've changed the wrapper to return both arrays.